L63-core-call-credentials-debug-string.md

```diff
@@ -0,0 +1,121 @@
+implementation of the method to all credential types. The output string depends
+on the kind of credential that is being parsed. The virtual method in the base
+class will have a default implementation so that it does not break any other
+custom credentials.
+
+## Rationale
+
+This provides a way to the user to learn more about the credentials they have
+set for a RPC. It may help with debugging any credentials related issues while
+sending the RPC. The design will also try to make sure that most of the
+computation and memory allocations is done during the DebugString method so that
+there is no default performance hits on the RPC.
+
+## Implementation
+
+### CallCredentials and SecureCallCredentials
+CallCredentials is the primary abstract class. Here we will add a virtual
+DebugString() method that returns an empty string by default.
+
+SecureCallCredentials inherits CallCredentials and will have an implementation
+that relies on its underlying grpc_call_credentials structure to get the
+debug string contents
+
+![CallCredentials class hierarchy](L63_graphics/call_creds_class_hierarchy.png)
+
+### C Call credentials and its derived classes
+The base C struct grpc_call_credentials will provide a helper debug_string().
+This will again be a virtual class that returns an empty string by default.
+We will discuss how every derived grpc_call_credentials struct will implement
+the debug_string method.  For now we will exclude grpc_plugin_credentials and
+grpc_composite_call_credentials. They will be discussed in separate sections
+
+![C CallCredentials hierarchy](L63_graphics/c_call_creds_hierarchy.png)
+
+### GRPC Oauth2 Token fetcher credentials (grpc_oauth2_token_fetcher_credentials)
+We could print whether a token is present and the token expiration time here.
+
+### GRPC Google IAM credentials (grpc_google_iam_credentials)
+We could print whether a token is present and the authority selector here.
+GRPC Service Account JWT Credentials (grpc_service_account_jwt_access_credentials)
+We could print the client id and token expiration data
+
+### GRPC Access Token Credentials (grpc_access_token_credentials)
+TBD
+
+
+### GRPC Plugin Credentials (grpc_plugin_credentials)
+The grpc plugin credentials, as the name suggests, uses plugins to extract
+auth metadata from a different kind of credentials.
+The extraction and other actions are achieved by maintaining callbacks in the
+plugin. To support a debug string we will similarly add callbacks to return a
+debug string and then the underlying plugin will determine how to populate
+that string.
+
+To do this we need demonstrate the flow of how a callback is invoked by the
+plugin
+
+
+![GRPC metadata creds plugin codeflow](L63_graphics/plugin_creds_codeflow.png)
+
+* The grpc_plugin_credentials composes a grpc_metadata_credentials_plugin
+that has all the callbacks to interact with the Credential type.
+* We add a grpc_metadata_credentials::(*debug_string)(void *state)
```

Comment:
Minor: use backticks to render this a code? (Also avoids issues with large spans being rendered oblique.)

Reply:
Done

Reply:
Done